tests/healthcheck.yaml

```
@@ -0,0 +1,117 @@
+version: "3.8"
+services:
+  autoheal:
+    image: willfarrell/autoheal
+    restart: unless-stopped
```

I think you need to add `privileged` permissions here. It does not seem to be a problem in the Github Actions runner, but I can't run the tests locally in my Fedora system without this, and it is also an issue we ran in other projects, like https://github.com/Tecnativa/docker-socket-proxy#usage
```suggestion
    restart: unless-stopped
    privileged: true
```

---

what's the error you are getting on fedora? (runs fine on github, gitlab, debian and ubuntu)

---

The container actually gives no error (maybe lack of logs from the `autoheal` docker container), it just exits with a non-zero exit code. Because of the `restart: unless-stopped`, it enters a loop where it is constantly restarting and can't actually "heal" other containers.

---

> runs fine on github, gitlab, debian and ubuntu

Those don't use SELinux, which is probably the one in charge of blocking such dangerous operations for unprivileged containers in Fedora.

---

Do you have some information from SELinux what actions are not permitted? (i would prefer to add seperate capabilities accoriding to https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities instead of giving the container all of them)

---

I couldn't find any useful information on the problem (except `--privileged`) so i added SELinux detection to run autoheal in privileged mode if needed.

@joao-p-marques can you test if the code is working for you (it tries to locate `getenforce`)?

---

I think you shouldn't care about all that, just add `--privileged`, because:

1. This is just a test, so we don't need to think on security so much as on a production instance for example.
2. If you're _not_ using SELinux and you're bind-mounting the docker socket on a container, you're basically giving it `--privileged` even if you didn't know it. That's the (in)security model of Docker itself.
3. Without bind-mounting the socket, the container cannot operate. With it, it has all privileges anyway.

If we were in production, you could workaround the problem with https://github.com/Tecnativa/docker-socket-proxy, which is built for that purpose. The socket proxy would be the privileged buddy, and other containers would be limited by it. That'd be a good fix here too, but... well, point 1 again. 😄

---

Well i think CI systems are prod systems as well. If one can inject code in other builds that would be bad. So i like to run my jobs with the least privileges possible as i think that's a good principle regardless of if its a prod system or not.

I'm aware that docker currently is giving many permissions to running docker containers (otherwise i wouldn't have discussed it would i?) that's a bad reason to give it more permissions.

With the current code we can run the tests with privileged where needed and without privileged where it's not needed.

---

I think what @Yajo was trying to say with point 2 is that not giving `privileged` in systems without `SELinux` is basically the same as giving them in systems that have, so the practical security impact would be the same, with or without it.

As stated in https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities, `privileged` should be always needed to do what we are doing here, so if it is not currently needed in some systems that might be the actual security compromise here.

Anyway, I have tested with your latest change and everything works as expected, so any of the 2 is fine by me :+1:

---

Exactly.

Just think about this: if you have raw access to the docker socket, what forbids you from starting a malicious privileged container on that socket? 🤷🏼‍♂️

Not adding `--privileged` for systems that don't use SELinux and still mounting the docker socket in the container is a placebo.

---

What about this?

---

Well at first the tests did not work on Fedora. I fixed that. @yajo did not fully agree with the solution but still approved the PR. @joao-p-marques did find it fine but never changed his approval to approved. For me this discussion for me was in the end mainly about https://en.wikipedia.org/wiki/Principle_of_least_privilege which i try to adhere to, so if I'm running tests on our CI they should run with the least possible privileges (which these tests ensure). And then the PR went nowhere for (almost) 3 years because of this (for me meta-)discussion.

---

OK, I trust your judgment then, as this is a bit out of my scope, but I want to push on the contributions.